### PR TITLE
Native auth MFA, AB#2083675, Fixes AB#2083675

### PR DESCRIPTION
**Pull request by SammyO**
Support for email OTP based MFA in native authentication flows.

- Added MFA results and states to the SDK interface
- Business logic & API integration
- Custom error handling for MFA required error during calls to /token with grant_type=refresh_token
- Private preview warnings in console and documentation
- Unit and E2E production tests

MSAL PR: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/2167

[AB#2083675](https://identitydivision.visualstudio.com/fac9d424-53d2-45c0-91b5-ef6ba7a6bf26/_workitems/edit/2083675)

### CODE REVIEW

**Review by fadidurah — APPROVED** [review]

**Review by Yuki-YuXin — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/controllers/results/MFACommandResult.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        override val correlationId: String,
+        val continuationToken: String,
+        val challengeTargetLabel: String,
+        val challengeChannel: String,
+        val codeLength: Int
+    ) : MFAChallengeCommandResult {
+        override fun toUnsanitizedString(): String = "VerificationRequired(correlationId=$correlationId, codeLength=$codeLength, challengeTargetLabel=$challengeTargetLabel, challengeChannel=$challengeChannel)"
+
+        override fun toString(): String = "VerificationRequired(correlationId=$correlationId, codeLength=$codeLength, challengeChannel=$challengeChannel)"
+    }
```
Comment (Yuki-YuXin) [inline common4j/src/main/com/microsoft/identity/common/java/nativeauth/controllers/results/MFACommandResult.kt:47]: Maybe challengeChannel is sensitive?

**Review by fadidurah — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java [hunk 2]
```diff
@@ -250,13 +263,9 @@ public static ServiceException getExceptionFromTokenErrorResponse(@NonNull final
     public static ServiceException convertToNativeAuthException(@NonNull final ServiceException exception) {
         final ServiceException outErr;
 
-        String message = "Token request failed.\nOriginal exception details: " + exception.getMessage();
-
-        // UiRequiredException is not a native authentication concept, so we convert it into a generic
-        // ServiceException
         outErr = new ServiceException(
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java:266]: Is this still needed?

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFAChallengeCommandParameters.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * The continuation token obtained from the token endpoint.
+     */
+    @NonNull
+    public final String continuationToken;
+
+    @NonNull
+    @Override
+    public String toUnsanitizedString() {
+        return "MFAChallengeCommandParameters(authority=" + authority + ", challengeTypes=" + challengeType + ")";
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFAChallengeCommandParameters.java]: is authority PII?

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFASelectedChallengeCommandParameters.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * The ID of the auth method that should be challenged.
+     */
+    @NonNull
+    public final String authMethodId;
+
+    @NonNull
+    @Override
+    public String toUnsanitizedString() {
+        return "MFASelectedChallengeCommandParameters(authority=" + authority + ", challengeTypes=" + challengeType + ", authMethodId=" + authMethodId + ")";
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFASelectedChallengeCommandParameters.java]: `challengeType=`?

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFASubmitChallengeCommandParameters.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * The continuation token obtained from the previous endpoint.
+     */
+    @NonNull
+    public final String continuationToken;
+
+    @NonNull
+    @Override
+    public String toUnsanitizedString() {
+        return "MFASubmitChallengeCommandParameters(authority=" + authority + ", challengeTypes=" + challengeType + ")";
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFASubmitChallengeCommandParameters.java]: `challengeType=`?

**Review by rmccahill — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/controllers/results/MFACommandResult.kt (hunk 1 above)
Comment (SammyO) [reply common4j/src/main/com/microsoft/identity/common/java/nativeauth/controllers/results/MFACommandResult.kt:47]: I don't think so. `challengeChannel` is `email`, `phone`, etc.

File: common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt [hunk 6]
```diff
@@ -319,6 +328,75 @@ class NativeAuthMsalController : BaseNativeAuthController() {
...
+                        errorCodes = tokenApiResult.errorCodes,
+                        correlationId = tokenApiResult.correlationId
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(
+                TAG,
+                parameters.correlationId,
+                "Exception thrown in signInSubmitCode",
```
Comment (SammyO) [inline common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt]: wrong method quoted here

File: common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt [hunk 7]
```diff
@@ -436,6 +512,145 @@ class NativeAuthMsalController : BaseNativeAuthController() {
...
+                        errorCodes = result.errorCodes,
+                        correlationId = result.correlationId
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(
+                TAG,
+                parameters.correlationId,
+                "Exception thrown in challengeDefaultAuth()",
```
Comment (SammyO) [inline common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt]: wrong method quoted

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFADefaultChallengeCommandParameters.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    /**
+     * The continuation token obtained from the token endpoint.
+     */
+    @NonNull
+    public final String continuationToken;
+
+    @NonNull
+    @Override
+    public String toUnsanitizedString() {
+        return "MFAChallengeCommandParameters(authority=" + authority + ", challengeTypes=" + challengeType + ")";
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/nativeauth/commands/parameters/MFADefaultChallengeCommandParameters.java]: class name wrongly quoted

File: common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt [hunk 9]
```diff
@@ -319,6 +328,75 @@ class NativeAuthMsalController : BaseNativeAuthController() {
...
+    fun signInSubmitChallenge(parameters: MFASubmitChallengeCommandParameters): MFASubmitChallengeCommandResult {
+        LogSession.logMethodCall(
+            tag = TAG,
+            correlationId = parameters.getCorrelationId(),
+            methodName = "${TAG}.signInSubmitChallenge"
+        )
+
+        try {
+            // Add default scopes
+            val mergedScopes: List<String> = addDefaultScopes(parameters.scopes)
```
Comment (nilo-ms) [inline common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt:340]: interesting, on iOS we add the default scopes only during the first "signIn(username, password: optional)" SDK call. Then the scopes are stored internally in memory (in the state). Why do you add the default scopes here too?

File: common/src/main/java/com/microsoft/identity/common/nativeauth/internal/util/CommandUtil.java [hunk 10]
```diff
@@ -142,4 +143,54 @@ public static SignInSubmitPasswordCommandParameters createSignInSubmitPasswordCo
...
+    ) {
+        return parameters.toBuilder()
+                .scopes(defaultScopes)
+                .correlationId(parameters.getCorrelationId())
+                .build();
+    }
+
+    /**
+     * Adds continuation token to [SignInStartCommandParameters] object and returns a new
+     * [SignInSubmitPasswordCommandParameters] object.
```
Comment (nilo-ms) [inline common/src/main/java/com/microsoft/identity/common/nativeauth/internal/util/CommandUtil.java]: nit: Should this "SignInSubmitCodeCommandParameters" ?

**Review by nilo-ms — APPROVED** [review]

File: common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt (hunk 9 above)
Comment (SammyO) [reply common/src/main/java/com/microsoft/identity/common/nativeauth/internal/controllers/NativeAuthMsalController.kt:340]: That would indeed be a better approach. Doing that is quite a refactoring in Android, as we pass scopes from one state to the next only if they're set by the developer (see interface classes, e.g. `MFAStates`. We don't take scopes from the result of an API call; but moreover, there is no result of an API call that includes scopes until the final token call is successful.
In signInSubmitCode we currently also set default scopes.

File: common/src/main/java/com/microsoft/identity/common/nativeauth/internal/util/CommandUtil.java (hunk 10 above)
Comment (SammyO) [reply common/src/main/java/com/microsoft/identity/common/nativeauth/internal/util/CommandUtil.java]: No, actually this entire comment is wrong.

**Pull request merged**